### ISSUE TITLE
Unable to yum localinstall on CentOS 7 according to docs

### ISSUE THREAD

**Issue by weikinhuang** [docker/docker#14042]
I am attempting to install `https://get.docker.com/rpm/1.7.0/centos-7/RPMS/x86_64/docker-engine-1.7.0-1.el7.centos.x86_64.rpm` according to the docs at [docs.docker.com](https://docs.docker.com/docker/installation/centos/) using `yum localinstall docker-engine-1.7.0-1.el7.centos.x86_64.rpm` on a freshly build CentOS7 (64bit) box, but it is unable to resolve certain dependencies.

`docker version`: Attempting to install 1.7.0

`docker info`: N/A

`uname -a`: `Linux build03.weikinhuang.com 3.10.0-229.4.2.el7.x86_64 #1 SMP Wed May 13 10:06:09 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`

`cat /etc/centos-release`: `CentOS Linux release 7.1.1503 (Core)`

Environment details (AWS, VirtualBox, physical, etc.):
Physical

How reproducible:
Attempted on several CentOS 7 machines. All the same issues

Steps to Reproduce:
1. `curl -O -sSL https://get.docker.com/rpm/1.7.0/centos-7/RPMS/x86_64/docker-engine-1.7.0-1.el7.centos.x86_64.rpm`
2. `yum localinstall docker-engine-1.7.0-1.el7.centos.x86_64.rpm`
3.

Actual Results:
```
yum localinstall docker-engine-1.7.0-1.el7.centos.x86_64.rpm 
Loaded plugins: fastestmirror
Examining docker-engine-1.7.0-1.el7.centos.x86_64.rpm: docker-engine-1.7.0-1.el7.centos.x86_64
Marking docker-engine-1.7.0-1.el7.centos.x86_64.rpm to be installed
Resolving Dependencies
--> Running transaction check
---> Package docker-engine.x86_64 0:1.7.0-1.el7.centos will be installed
--> Processing Dependency: libc.so.6 for package: docker-engine-1.7.0-1.el7.centos.x86_64
Loading mirror speeds from cached hostfile
 * base: mirrors.centos.webair.com
 * epel: mirror.us.leaseweb.net
 * extras: mirror.vcu.edu
 * updates: mirror.netdepot.com
--> Processing Dependency: libpthread.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libsqlite3.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Finished Dependency Resolution
Error: Package: docker-engine-1.7.0-1.el7.centos.x86_64 (/docker-engine-1.7.0-1.el7.centos.x86_64)
           Requires: libpthread.so.0
Error: Package: docker-engine-1.7.0-1.el7.centos.x86_64 (/docker-engine-1.7.0-1.el7.centos.x86_64)
           Requires: libsqlite3.so.0
Error: Package: docker-engine-1.7.0-1.el7.centos.x86_64 (/docker-engine-1.7.0-1.el7.centos.x86_64)
           Requires: libc.so.6
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```

```
yum localinstall docker-engine-1.7.0-1.el7.centos.x86_64.rpm --skip-broken
Loaded plugins: fastestmirror
Examining docker-engine-1.7.0-1.el7.centos.x86_64.rpm: docker-engine-1.7.0-1.el7.centos.x86_64
Marking docker-engine-1.7.0-1.el7.centos.x86_64.rpm to be installed
Resolving Dependencies
--> Running transaction check
---> Package docker-engine.x86_64 0:1.7.0-1.el7.centos will be installed
--> Processing Dependency: libc.so.6 for package: docker-engine-1.7.0-1.el7.centos.x86_64
Loading mirror speeds from cached hostfile
 * base: mirrors.centos.webair.com
 * epel: mirror.us.leaseweb.net
 * extras: mirror.vcu.edu
 * updates: mirror.netdepot.com
--> Processing Dependency: libpthread.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libsqlite3.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64

Packages skipped because of dependency problems:
    docker-engine-1.7.0-1.el7.centos.x86_64 from /docker-engine-1.7.0-1.el7.centos.x86_64
```

Expected Results:
Expects to use existing already installed dependencies.

Additional info:
```
[root@docker003 ~]$ yum provides '*/libpthread.so.0'
glibc-2.17-78.el7.x86_64 : The GNU libc libraries
Repo        : @base
Matched from:
Filename    : /lib64/libpthread.so.0

[root@docker003 ~]$ yum provides '*/libc.so.6'
glibc-2.17-78.el7.x86_64 : The GNU libc libraries
Repo        : @base
Matched from:
Filename    : /lib64/libc.so.6

[root@docker003 ~]$ yum install glibc
Package glibc-2.17-78.el7.x86_64 already installed and latest version
Nothing to do

[root@docker003 ~]$ yum provides '*/libsqlite3.so.0'
sqlite-3.7.17-4.el7.x86_64 : Library that implements an embeddable SQL database engine
Repo        : @anaconda
Matched from:
Filename    : /usr/lib64/libsqlite3.so.0

[root@docker003 ~]$ yum install sqlite
Package sqlite-3.7.17-4.el7.x86_64 already installed and latest version
Nothing to do
```

**Comment by henrysher** [comment]
@weikinhuang could u plz show your yum conf and enabled yum repos? I have verified it in my fresh-installed centos7, no such problem, it works well as shown in below. It seems docker-engine rpm needs the dependencies from i686 packages like glibc, libcgroup, libgcc, libstdc++, sqilte, etc.

```
[root@docker-host /]# cat /etc/redhat-release 
CentOS Linux release 7.1.1503 (Core) 
[root@docker-host /]# curl -O -sSL https://get.docker.com/rpm/1.7.0/centos-7/RPMS/x86_64/docker-engine-1.7.0-1.el7.centos.x86_64.rpm
[root@docker-host /]# yum localinstall docker-engine-1.7.0-1.el7.centos.x86_64.rpm
Loaded plugins: fastestmirror
Examining docker-engine-1.7.0-1.el7.centos.x86_64.rpm: docker-engine-1.7.0-1.el7.centos.x86_64
Marking docker-engine-1.7.0-1.el7.centos.x86_64.rpm to be installed
Resolving Dependencies
--> Running transaction check
---> Package docker-engine.x86_64 0:1.7.0-1.el7.centos will be installed
--> Processing Dependency: libc.so.6 for package: docker-engine-1.7.0-1.el7.centos.x86_64
Loading mirror speeds from cached hostfile
 * base: mirrors.psychz.net
 * extras: mirrors.psychz.net
 * updates: mirror.san.fastserv.com
--> Processing Dependency: libcgroup for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libpthread.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libsqlite3.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: tar for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Running transaction check
---> Package glibc.i686 0:2.17-78.el7 will be installed
--> Processing Dependency: libfreebl3.so(NSSRAWHASH_3.12.3) for package: glibc-2.17-78.el7.i686
--> Processing Dependency: libfreebl3.so for package: glibc-2.17-78.el7.i686
---> Package libcgroup.x86_64 0:0.41-8.el7 will be installed
---> Package sqlite.i686 0:3.7.17-4.el7 will be installed
--> Processing Dependency: libtinfo.so.5 for package: sqlite-3.7.17-4.el7.i686
--> Processing Dependency: libreadline.so.6 for package: sqlite-3.7.17-4.el7.i686
--> Processing Dependency: libncurses.so.5 for package: sqlite-3.7.17-4.el7.i686
---> Package tar.x86_64 2:1.26-29.el7 will be installed
--> Running transaction check
---> Package ncurses-libs.i686 0:5.9-13.20130511.el7 will be installed
--> Processing Dependency: libstdc++.so.6(GLIBCXX_3.4) for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libstdc++.so.6(CXXABI_1.3) for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libstdc++.so.6 for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libgcc_s.so.1(GCC_3.0) for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libgcc_s.so.1 for package: ncurses-libs-5.9-13.20130511.el7.i686
---> Package nss-softokn-freebl.i686 0:3.16.2.3-9.el7 will be installed
---> Package readline.i686 0:6.2-9.el7 will be installed
--> Running transaction check
---> Package libgcc.i686 0:4.8.3-9.el7 will be installed
---> Package libstdc++.i686 0:4.8.3-9.el7 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

====================================================================================================================
 Package                 Arch        Version                    Repository                                     Size
====================================================================================================================
Installing:
 docker-engine           x86_64      1.7.0-1.el7.centos         /docker-engine-1.7.0-1.el7.centos.x86_64       19 M
Installing for dependencies:
 glibc                   i686        2.17-78.el7                base                                          4.2 M
 libcgroup               x86_64      0.41-8.el7                 base                                           64 k
 libgcc                  i686        4.8.3-9.el7                base                                           99 k
 libstdc++               i686        4.8.3-9.el7                base                                          307 k
 ncurses-libs            i686        5.9-13.20130511.el7        base                                          316 k
 nss-softokn-freebl      i686        3.16.2.3-9.el7             base                                          187 k
 readline                i686        6.2-9.el7                  base                                          189 k
 sqlite                  i686        3.7.17-4.el7               base                                          396 k
 tar                     x86_64      2:1.26-29.el7              base                                          843 k

Transaction Summary
====================================================================================================================
Install  1 Package (+9 Dependent packages)

Total size: 26 M
Total download size: 6.5 M
Installed size: 40 M
Is this ok [y/d/N]: y
Downloading packages:
(1/9): libcgroup-0.41-8.el7.x86_64.rpm                                                       |  64 kB  00:00:05     
(2/9): libgcc-4.8.3-9.el7.i686.rpm                                                           |  99 kB  00:00:05     
(3/9): nss-softokn-freebl-3.16.2.3-9.el7.i686.rpm                                            | 187 kB  00:00:00     
(4/9): readline-6.2-9.el7.i686.rpm                                                           | 189 kB  00:00:00     
(5/9): ncurses-libs-5.9-13.20130511.el7.i686.rpm                                             | 316 kB  00:00:06     
(6/9): sqlite-3.7.17-4.el7.i686.rpm                                                          | 396 kB  00:00:00     
(7/9): libstdc++-4.8.3-9.el7.i686.rpm                                                        | 307 kB  00:00:06     
(8/9): tar-1.26-29.el7.x86_64.rpm                                                            | 843 kB  00:00:02     
(9/9): glibc-2.17-78.el7.i686.rpm                                                            | 4.2 MB  00:00:08     
--------------------------------------------------------------------------------------------------------------------
Total                                                                               769 kB/s | 6.5 MB  00:00:08     
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : libgcc-4.8.3-9.el7.i686                                                                         1/10 
  Installing : nss-softokn-freebl-3.16.2.3-9.el7.i686                                                          2/10 
  Installing : glibc-2.17-78.el7.i686                                                                          3/10 
  Installing : 2:tar-1.26-29.el7.x86_64                                                                        4/10 
  Installing : libcgroup-0.41-8.el7.x86_64                                                                     5/10 
  Installing : libstdc++-4.8.3-9.el7.i686                                                                      6/10 
  Installing : ncurses-libs-5.9-13.20130511.el7.i686                                                           7/10 
  Installing : readline-6.2-9.el7.i686                                                                         8/10 
  Installing : sqlite-3.7.17-4.el7.i686                                                                        9/10 
  Installing : docker-engine-1.7.0-1.el7.centos.x86_64                                                        10/10 
  Verifying  : 2:tar-1.26-29.el7.x86_64                                                                        1/10 
  Verifying  : libgcc-4.8.3-9.el7.i686                                                                         2/10 
  Verifying  : libcgroup-0.41-8.el7.x86_64                                                                     3/10 
  Verifying  : docker-engine-1.7.0-1.el7.centos.x86_64                                                         4/10 
  Verifying  : sqlite-3.7.17-4.el7.i686                                                                        5/10 
  Verifying  : ncurses-libs-5.9-13.20130511.el7.i686                                                           6/10 
  Verifying  : libstdc++-4.8.3-9.el7.i686                                                                      7/10 
  Verifying  : glibc-2.17-78.el7.i686                                                                          8/10 
  Verifying  : readline-6.2-9.el7.i686                                                                         9/10 
  Verifying  : nss-softokn-freebl-3.16.2.3-9.el7.i686                                                         10/10 

Installed:
  docker-engine.x86_64 0:1.7.0-1.el7.centos                                                                         

Dependency Installed:
  glibc.i686 0:2.17-78.el7      libcgroup.x86_64 0:0.41-8.el7            libgcc.i686 0:4.8.3-9.el7                
  libstdc++.i686 0:4.8.3-9.el7  ncurses-libs.i686 0:5.9-13.20130511.el7  nss-softokn-freebl.i686 0:3.16.2.3-9.el7 
  readline.i686 0:6.2-9.el7     sqlite.i686 0:3.7.17-4.el7               tar.x86_64 2:1.26-29.el7                 

Complete!
```

**Comment by weikinhuang** [comment]
@henrysher Ah, why would this docker binary be compiled needing 32bit dependencies. I explicitly disable installing any `i*86` packages on my systems. My `yum.conf` contains `exclude=*.i386 *.i586 *.i686` which would explain my issue. But as a separate issue. The docker binaries/rpm should be compiled needing the 32bit dependencies.

I guess I will wait until 1.7.0 enters the extras repo in centos or perhaps 1.7.1 will be compiled properly.

**Comment by thresheek** [comment]
The RPM package has hard-coded Requires, which are obviously wrong for a 64-bit system:

https://github.com/docker/docker/blob/master/hack/make/.build-rpm/docker-engine.spec#L36

Compare that to, say, glibc:

# rpm -q --provides glibc | grep libc.so.6
libc.so.6()(64bit)
libc.so.6(GLIBC_2.10)(64bit)
libc.so.6(GLIBC_2.11)(64bit)
libc.so.6(GLIBC_2.12)(64bit)
libc.so.6(GLIBC_2.2.5)(64bit)
libc.so.6(GLIBC_2.2.6)(64bit)
libc.so.6(GLIBC_2.3)(64bit)
libc.so.6(GLIBC_2.3.2)(64bit)
libc.so.6(GLIBC_2.3.3)(64bit)
libc.so.6(GLIBC_2.3.4)(64bit)
libc.so.6(GLIBC_2.4)(64bit)
libc.so.6(GLIBC_2.5)(64bit)
libc.so.6(GLIBC_2.6)(64bit)
libc.so.6(GLIBC_2.7)(64bit)
libc.so.6(GLIBC_2.8)(64bit)
libc.so.6(GLIBC_2.9)(64bit)

**Comment by salimane** [comment]
:+1:

**Comment by houqianming** [comment]
@weikinhuang @thresheek I got the same error,  and there is no exclude config in my yum.conf，how to work around this problem?

**Comment by LK4D4** [comment]
Dunno who to ping, so I ping @jfrazelle and @vbatts :)

**Comment by Djelibeybi** [comment]
Same issue on the CentOS7 build, which is unsurprising given they use the same underlying specfile.

**Comment by jfrazelle** [comment]
ping @vbatts idk how this is happening

**Comment by jfrazelle** [comment]
I cannot reproduce

**Comment by Djelibeybi** [comment]
@jfrazelle interesting. Do you perhaps already have the i686 RPMs installed? It's 100% reproducible on a minimal install of Oracle Linux 7 using the current stable RPM (1.7.0).

**Comment by jfrazelle** [comment]
i have tried on all centos version and all fedora versions

**Comment by vbatts** [comment]
The tab is open in my browser, but I didn't investigate it today. It's on
the docket for tomorrow.

**Comment by Djelibeybi** [comment]
RPM AutoRequires should automatically pull in the necessary packages. I'm testing this theory now.

**Comment by weikinhuang** [comment]
Happens when attempting to install via centos:latest docker image as well
```bash
[root@docker003 ~]$ docker run -it --rm centos:latest bash
Unable to find image 'centos:latest' locally
latest: Pulling from centos

f1b10cd84249: Pull complete 
c852f6d61e65: Pull complete 
7322fbe74aa5: Already exists 
centos:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.

Digest: sha256:a4627c43bafc86705af2e8a5ea1f0ed34fbf27b6e7a392e5ee45dbd4736627cc
Status: Downloaded newer image for centos:latest
[root@docker.7fc7a4bd0635 /]♫ curl -O -sSL https://get.docker.com/rpm/1.7.0/centos-7/RPMS/x86_64/docker-engine-1.7.0-1.el7.centos.x86_64.rpm
[root@docker.7fc7a4bd0635 /]♫ yum localinstall docker-engine-1.7.0-1.el7.centos.x86_64.rpm
Loaded plugins: fastestmirror
Examining docker-engine-1.7.0-1.el7.centos.x86_64.rpm: docker-engine-1.7.0-1.el7.centos.x86_64
Marking docker-engine-1.7.0-1.el7.centos.x86_64.rpm to be installed
Resolving Dependencies
--> Running transaction check
---> Package docker-engine.x86_64 0:1.7.0-1.el7.centos will be installed
--> Processing Dependency: libc.so.6 for package: docker-engine-1.7.0-1.el7.centos.x86_64
base                                                                                                                              | 3.6 kB  00:00:00     
extras                                                                                                                            | 3.4 kB  00:00:00     
updates                                                                                                                           | 3.4 kB  00:00:00     
(1/4): base/7/x86_64/group_gz                                                                                                     | 154 kB  00:00:00     
(2/4): extras/7/x86_64/primary_db                                                                                                 |  62 kB  00:00:01     
(3/4): updates/7/x86_64/primary_db                                                                                                | 2.5 MB  00:00:06     
(4/4): base/7/x86_64/primary_db                                                                                                   | 5.1 MB  00:00:54     
Determining fastest mirrors
 * base: mirrors.centos.webair.com
 * extras: mirror.lug.udel.edu
 * updates: mirror.lug.udel.edu
--> Processing Dependency: libcgroup for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libdevmapper.so.1.02(Base)(64bit) for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libpthread.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libsqlite3.so.0 for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Processing Dependency: libdevmapper.so.1.02()(64bit) for package: docker-engine-1.7.0-1.el7.centos.x86_64
--> Running transaction check
---> Package device-mapper-libs.x86_64 7:1.02.93-3.el7_1.1 will be installed
--> Processing Dependency: device-mapper = 7:1.02.93-3.el7_1.1 for package: 7:device-mapper-libs-1.02.93-3.el7_1.1.x86_64
---> Package glibc.i686 0:2.17-78.el7 will be installed
--> Processing Dependency: libfreebl3.so(NSSRAWHASH_3.12.3) for package: glibc-2.17-78.el7.i686
--> Processing Dependency: libfreebl3.so for package: glibc-2.17-78.el7.i686
---> Package libcgroup.x86_64 0:0.41-8.el7 will be installed
---> Package sqlite.i686 0:3.7.17-4.el7 will be installed
--> Processing Dependency: libtinfo.so.5 for package: sqlite-3.7.17-4.el7.i686
--> Processing Dependency: libreadline.so.6 for package: sqlite-3.7.17-4.el7.i686
--> Processing Dependency: libncurses.so.5 for package: sqlite-3.7.17-4.el7.i686
--> Running transaction check
---> Package device-mapper.x86_64 7:1.02.93-3.el7_1.1 will be installed
---> Package ncurses-libs.i686 0:5.9-13.20130511.el7 will be installed
--> Processing Dependency: libstdc++.so.6(GLIBCXX_3.4) for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libstdc++.so.6(CXXABI_1.3) for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libstdc++.so.6 for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libgcc_s.so.1(GCC_3.0) for package: ncurses-libs-5.9-13.20130511.el7.i686
--> Processing Dependency: libgcc_s.so.1 for package: ncurses-libs-5.9-13.20130511.el7.i686
---> Package nss-softokn-freebl.i686 0:3.16.2.3-9.el7 will be installed
---> Package readline.i686 0:6.2-9.el7 will be installed
--> Running transaction check
---> Package libgcc.i686 0:4.8.3-9.el7 will be installed
---> Package libstdc++.i686 0:4.8.3-9.el7 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

=========================================================================================================================================================
 Package                          Arch                 Version                              Repository                                              Size
=========================================================================================================================================================
Installing:
 docker-engine                    x86_64               1.7.0-1.el7.centos                   /docker-engine-1.7.0-1.el7.centos.x86_64                19 M
Installing for dependencies:
 device-mapper                    x86_64               7:1.02.93-3.el7_1.1                  updates                                                209 k
 device-mapper-libs               x86_64               7:1.02.93-3.el7_1.1                  updates                                                259 k
 glibc                            i686                 2.17-78.el7                          base                                                   4.2 M
 libcgroup                        x86_64               0.41-8.el7                           base                                                    64 k
 libgcc                           i686                 4.8.3-9.el7                          base                                                    99 k
 libstdc++                        i686                 4.8.3-9.el7                          base                                                   307 k
 ncurses-libs                     i686                 5.9-13.20130511.el7                  base                                                   316 k
 nss-softokn-freebl               i686                 3.16.2.3-9.el7                       base                                                   187 k
 readline                         i686                 6.2-9.el7                            base                                                   189 k
 sqlite                           i686                 3.7.17-4.el7                         base                                                   396 k

Transaction Summary
=========================================================================================================================================================
Install  1 Package (+10 Dependent packages)

Total size: 25 M
Total download size: 6.2 M
Installed size: 38 M
Is this ok [y/d/N]: 
```

**Comment by jfrazelle** [comment]
what makes no sense are the deps we dont even specify...

**Comment by vbatts** [comment]
Ah. This looks like a meta-requires issue. The devicemapper libs are
requiring libc, which is fine and expected, _but_ why it thinks the i686
libc satisfies this, will take a little investigation. Cc @lsm5 did you
have to deal with this in building the rpm spec for centos?

**Comment by Djelibeybi** [comment]
Those are the dependencies for the ones you do. The yum output is pretty clear on the dependency chain. I think I've got a fix (removing the specific dependency calls and letting rpmbuild auto-detect the dependencies), but I need time to let all the builds run to test the output.

**Comment by Djelibeybi** [comment]
Ok, have fixed it for C7/OL7. Now I need to run the same fix across all the RPM distros to make sure it still works on all of them. See you in a few hours. :)

**Comment by lsm5** [comment]
looking at fedora spec files, (lines 38-43)[https://github.com/docker/docker/blob/master/hack/make/.build-rpm/docker-engine.spec#L38-L43] are probably not required, maybe try building without those? i'll look deeper and post if i see anything else.

**Comment by lsm5** [comment]
looks like @Djelibeybi has a PR going for that already, nvm me :)

**Comment by jfrazelle** [comment]
fixed by #14456
